Add possibility to use unix domain socket instead of a normal port

mortbauer opened the pull request:
If you supply the option --socket-path the proxy server will use the supplied path to create a unix domain socket instead of the localhost port.

This has many advantages in my opinion.

Best Regards Martin

reubenmiller commented on c8ylp/cli/core.py:
```diff
@@ -633,12 +634,20 @@ def start_proxy(
-            opts.tcp_timeout,
-        )
+        if opts.socket_path is not None:
+            tcp_server = UnixStreamProxyServer(
```
We should change the variable name from "tcp_server" to something more generic, like "server" since we now support different types of servers

mortbauer replied:
Ok, I did a few more commits to integrate your wishes,

* did the linting, formatting
* added a very basic unit test regarding the cli, didn't find any "real" test for the server except in the integration tests
* I customized the usage example on the cli if invoked with the socket-path option and added a small example to the readme

mortbauer commented:
Ok, I did a few more commits to integrate your wishes,

* did the linting, formatting
* added a very basic unit test regarding the cli, didn't find any "real" test for the server except in the integration tests
* I customized the usage example on the cli if invoked with the socket-path option and added a small example to the readme

switschel commented on c8ylp/cli/core.py:
```diff
@@ -629,32 +630,40 @@ def start_proxy(
-            opts.tcp_timeout,
-        )
+        if opts.socket_path is not None:
+            server = UnixStreamProxyServer(
```
This will not work in non-unix OS. We should handle this and don't allow "--socket-path" on non-unix OS

mortbauer replied:
renamed it to socket_server

mortbauer commented on c8ylp/options.py:
```diff
@@ -243,6 +243,15 @@ def deprecated(ctx: click.Context, _param, value) -> Any:
     help="TFA Code. Required when the 'TFA enabled' is enabled for a user",
 )
 
+SOCKET_PATH = click.option(
```
ok I made this now platform specific, so the option is only available if the AF_UNIX socket type is available

switschel commented on c8ylp/cli/core.py:
```diff
@@ -35,7 +35,7 @@
 from ..env import save_env
 from ..rest_client.c8yclient import CumulocityClient, CumulocityMissingTFAToken
-from ..tcp_socket import TCPProxyServer
+from ..tcp_socket import TCPProxyServer, UnixStreamProxyServer
```
This import makes the Windows Unit Test fail:
c8ylp\cli\core.py:38: in <module>
    from ..tcp_socket import TCPProxyServer, UnixStreamProxyServer
c8ylp\tcp_socket\__init__.py:20: in <module>
    from c8ylp.tcp_socket.tcp_server import TCPProxyServer, UnixStreamProxyServer
c8ylp\tcp_socket\tcp_server.py:123: in <module>
    class CustomUnixStreamServer(socketserver.UnixStreamServer):
E   AttributeError: module 'socketserver' has no attribute 'UnixStreamServer'

We can check the socket or the OS and make a conditional import
```suggestion
import platform
if 'Linux' == platform.system():
    from ..tcp_socket import UnixStreamProxyServer
from ..tcp_socket import TCPProxyServer
```

mortbauer replied:
Refactored, the unixstreamproxyserver lives in its own module now, and the imports are check the AF_UNIX attribute of the socket module, this is what they also do in the socketserver module.

switschel commented:
@reubenmiller Do you have an idea why the Unit Test for Windows again failed?

reubenmiller commented:
@switschel The problem was caused by a None comparison (instead of a boolean / falsey) which then caused the non-existant unix library to be erroneously imported.

reubenmiller commented:
@switschel, @mortbauer: Do you think it is a good idea that `--socket-path` is a dynamic property, this might make it hard for people to debug across Windows/Unix-like environments. It might be better to just put validation on the option, and throw an exception if it is being called on windows (or more specifically where AF_UNIX is not defined in the socket lib)

mortbauer commented:
@reubenmiller this changes definitely make it more obvious what is going on, so for me it looks great. 
I just have a question regarding github now, how can you push to merepository? or was this even you, or who is y509540?

reubenmiller commented:
> @reubenmiller this changes definitely make it more obvious what is going on, so for me it looks great.
> I just have a question regarding github now, how can you push to merepository? or was this even you, or who is y509540?

Yeah that was me. I think when a PR is created that it might be linked to the upstream repo (and not the source)…but would need to look it up myself

switschel commented:
> @switschel, @mortbauer: Do you think it is a good idea that `--socket-path` is a dynamic property, this might make it hard for people to debug across Windows/Unix-like environments. It might be better to just put validation on the option, and throw an exception if it is being called on windows (or more specifically where AF_UNIX is not defined in the socket lib)

You're absolutely right. It think it looks good now after your latest commit.
@reubenmiller Do you trigger the merge of this PR?

reubenmiller commented:
@mortbauer Thanks again for your PR. The changes are included in the latest release v2.2.0
